File: conversion/layers/resource/builder.py
# @author: Davy Bellens

"""Orchestrator for Resource layer solid generation."""

import conversion.layers.resource.appearance as appearance
from conversion.layers.resource.math3d import (
    IDENTITY, cross, dot, mat_multiply, normalize, scale, subtract
)
from conversion.layers.resource.mesh import Mesh
from conversion.layers.resource.placement import axis_placement_matrix, read_direction, read_point
from conversion.layers.resource.solid import (
    extruded_area_solid, faceted_brep, polygonal_face_set, surface_model, triangulated_face_set
)

SOLID_BUILDERS = {
    "IfcExtrudedAreaSolid": extruded_area_solid,
    "IfcFacetedBrep": faceted_brep,
    "IfcTriangulatedFaceSet": triangulated_face_set,
    "IfcPolygonalFaceSet": polygonal_face_set,
    "IfcFaceBasedSurfaceModel": surface_model,
    "IfcShellBasedSurfaceModel": surface_model,
}


class UnsupportedGeometry(Exception):
    """Raised when a Resource-layer item has no builder."""


def build_item(item, styles=True):
    mesh = _build_geometry(item, styles)
    if styles:
        mesh.fill_style(appearance.item_rgba(item))
    return mesh


def _build_geometry(item, styles=True):
    name = item.is_a()

    builder = SOLID_BUILDERS.get(name)
    if builder is not None:
        return builder(item)

    if name in ("IfcBooleanResult", "IfcBooleanClippingResult"):
        return build_item(item.FirstOperand, styles)

    if name == "IfcMappedItem":
        source = item.MappingSource
        origin = axis_placement_matrix(source.MappingOrigin)
        target = _transformation_operator_matrix(item.MappingTarget)
        mesh = Mesh()
        for sub in source.MappedRepresentation.Items:
            try:
                mesh.extend(build_item(sub, styles))
            except UnsupportedGeometry:
                continue
        return mesh.transformed(mat_multiply(target, origin))

    raise UnsupportedGeometry(name)


def _transformation_operator_matrix(op):
    if op is None:
        return IDENTITY
    origin = read_point(op.LocalOrigin)

    non_uniform = op.is_a("IfcCartesianTransformationOperator3DnonUniform")
    scale1 = float(getattr(op, "Scale", None) or 1.0)
    scale2 = float(getattr(op, "Scale2", None) or scale1) if non_uniform else scale1
    scale3 = float(getattr(op, "Scale3", None) or scale1) if non_uniform else scale1

    z_axis = normalize(read_direction(getattr(op, "Axis3", None), (0.0, 0.0, 1.0)))
    x_axis = normalize(read_direction(getattr(op, "Axis1", None), (1.0, 0.0, 0.0)))
    projected = subtract(x_axis, scale(z_axis, dot(x_axis, z_axis)))
    if dot(projected, projected) < 1e-20:
        projected = (1.0, 0.0, 0.0)
    x_axis = normalize(projected)

    # When Axis2 is actually supplied, take it as authored rather than
    # re-deriving it from X and Z. A mirrored mapping target (left/right-hand
    # instance) is commonly expressed by an Axis2 that is *not* simply
    # cross(Z, X); deriving it instead silently un-mirrors the instance.
    axis2 = getattr(op, "Axis2", None)
    if axis2 is not None:
        y_axis = normalize(read_direction(axis2, (0.0, 1.0, 0.0)))
    else:
        y_axis = cross(z_axis, x_axis)

    return (
        (x_axis[0] * scale1, y_axis[0] * scale2, z_axis[0] * scale3, origin[0]),
        (x_axis[1] * scale1, y_axis[1] * scale2, z_axis[1] * scale3, origin[1]),
        (x_axis[2] * scale1, y_axis[2] * scale2, z_axis[2] * scale3, origin[2]),
        (0.0, 0.0, 0.0, 1.0),
    )

File: UnityPackage/com.byvad.ifc-import/Runtime/Layers/Resource/Builder.cs
// @author: Davy Bellens

using System;
using System.Collections.Generic;
using Conversion.Ifc;

namespace Conversion.Layers.Resource
{
    /// <summary>Raised when a Resource-layer item has no builder.</summary>
    public sealed class UnsupportedGeometryException : Exception
    {
        public UnsupportedGeometryException(string message) : base(message)
        {
        }
    }

    /// <summary>Orchestrator for Resource-layer solid generation.</summary>
    public sealed class Builder
    {
        private readonly HoleStats _stats;
        private readonly Appearance _appearance;

        /// <summary>Guards against a representation map that maps itself.</summary>
        private const int MaxDepth = 16;

        public Builder(HoleStats stats = null, Appearance appearance = null)
        {
            _stats = stats ?? new HoleStats();
            _appearance = appearance ?? new Appearance();
        }

        public HoleStats Stats => _stats;

        /// <summary>
        /// Radians per unit of the project's plane-angle measure, or 0 to guess.
        /// Set this from Units.PlaneAngleScale before converting: IFC2X3 files are
        /// commonly authored in degrees, and an arc trim parameter cannot be read
        /// correctly without knowing which.
        /// </summary>
        public double PlaneAngleScale { get; set; }

        public Mesh BuildItem(IfcEntity item, bool styles) => BuildItem(item, styles, 0);

        private Mesh BuildItem(IfcEntity item, bool styles, int depth)
        {
            Mesh mesh = BuildGeometry(item, styles, depth);
            if (styles)
            {
                mesh.FillStyle(_appearance.ItemRgba(item));
            }
            return mesh;
        }

        private Mesh BuildGeometry(IfcEntity item, bool styles, int depth)
        {
            if (item == null)
            {
                throw new UnsupportedGeometryException("(null item)");
            }
            if (depth > MaxDepth)
            {
                throw new UnsupportedGeometryException($"{item.IsA()} (nested too deeply)");
            }

            string name = item.IsA();

            switch (name)
            {
                case "IfcExtrudedAreaSolid":
                    return Solid.ExtrudedAreaSolid(item, _stats, PlaneAngleScale);
                case "IfcFacetedBrep":
                    return Solid.FacetedBrep(item, _stats);
                case "IfcTriangulatedFaceSet":
                    return Solid.TriangulatedFaceSet(item);
                case "IfcPolygonalFaceSet":
                    return Solid.PolygonalFaceSet(item, _stats);
                case "IfcFaceBasedSurfaceModel":
                case "IfcShellBasedSurfaceModel":
                    return Solid.SurfaceModel(item, _stats);
            }

            if (item.IsA("IfcBooleanResult"))
            {
                return BooleanResult(item, styles, depth);
            }

            if (name == "IfcMappedItem")
            {
                return MappedItem(item, styles, depth);
            }

            throw new UnsupportedGeometryException(name);
        }

        /// <summary>
        /// IfcBooleanResult and its clipping subtype.
        /// <para>
        /// The old behaviour of returning FirstOperand untouched is kept only as a
        /// fallback: a clipped solid drawn uncut is wrong, but it is a great deal
        /// less wrong than nothing at all.
        /// </para>
        /// </summary>
        private Mesh BooleanResult(IfcEntity item, bool styles, int depth)
        {
            Mesh first = BuildItem(item.Entity("FirstOperand"), styles, depth + 1);
            if (first.IsEmpty)
            {
                return first;
            }

            IfcEntity secondEntity = item.Entity("SecondOperand");
            if (secondEntity == null)
            {
                return first;
            }

            Mesh second = ResolveSecondOperand(secondEntity, first, depth);
            if (second == null || second.IsEmpty)
            {
                return first;
            }

            return ApplyOperator(item.String("Operator"), first, second);
        }

        private Mesh ResolveSecondOperand(IfcEntity secondEntity, Mesh first, int depth)
        {
            try
            {
                if (secondEntity.IsA("IfcHalfSpaceSolid"))
                {
                    Bounds(first, out Vec3 min, out Vec3 max);
                    return Solid.HalfSpaceSolid(secondEntity, min, max, _stats, PlaneAngleScale);
                }
                return BuildItem(secondEntity, false, depth + 1);
            }
            catch (UnsupportedGeometryException)
            {
                return null;
            }
        }

        private static Mesh ApplyOperator(string op, Mesh first, Mesh second)
        {
            switch (op ?? "DIFFERENCE")
            {
                case "UNION":
                    return MeshBoolean.Union(first, second);
                case "INTERSECTION":
                    return MeshBoolean.Intersect(first, second);
                default:
                    return MeshBoolean.Subtract(first, new[] { second });
            }
        }

        private Mesh MappedItem(IfcEntity item, bool styles, int depth)
        {
            IfcEntity source = item.Entity("MappingSource");
            if (source == null)
            {
                throw new UnsupportedGeometryException("IfcMappedItem (no MappingSource)");
            }

            Matrix4 origin = Placement.AxisPlacementMatrix(source.Entity("MappingOrigin"));
            Matrix4 target = Placement.TransformationOperatorMatrix(item.Entity("MappingTarget"));

            var mesh = new Mesh();
            IfcEntity representation = source.Entity("MappedRepresentation");
            if (representation != null)
            {
                foreach (IfcEntity sub in representation.Entities("Items"))
                {
                    try
                    {
                        mesh.Extend(BuildItem(sub, styles, depth + 1));
                    }
                    catch (UnsupportedGeometryException)
                    {
                        // One unsupported item in a mapped block should not lose the rest.
                    }
                }
            }

            return mesh.Transformed(Matrix4.Multiply(target, origin));
        }

        internal static void Bounds(Mesh mesh, out Vec3 min, out Vec3 max)
        {
            var bounds = BoundsAccumulator.Empty;
            bounds.Add(mesh.Vertices);
            bounds.TryGetBounds(out min, out max);
        }
    }
}